### ISSUE TITLE
Compatibility break on 'hasOwnProperty' usage

### ISSUE THREAD

**Issue by hiredgun** [sylvainpolletvillard/ObjectModel#52]
The current version (3.1) introduced compatibility break for 'hasOwnProperty' method usage.

Code snippet to reproduce:
```
const { Model } = require('objectmodel');

class Foo extends Model({
  id: Number,
  status: String,
}).defaults({
  status: 'pending',
}) { }

const foo = new Foo({ id: 1 }); // init only with id

console.log(JSON.stringify(foo)); // {"id":1,"status":"pending"}
console.log(foo.hasOwnProperty('id'));  // true
console.log(foo.hasOwnProperty('status'));  // false  < ------ false even if default is specified - v2.6.2 returns true
```

**Comment by sylvainpolletvillard** [comment]
Hi,

It was actually a bug in v2 that I fixed in v3. Defaults are supposed to be values set on the model prototype, not on the instance itself, as explained in the docs:

> To specify default values for some properties of your object models, put them in the **model prototype**. You can also use the defaults method as a shorthand for setting all the default values at once.

So they are not ownProperties, but the `in` operator still works: `"status" in foo`

**Comment by hiredgun** [comment]
IMHO the previous approach was correct and the current implementation violates the core feature of this library. In "What is this library" section of your docs you say that this library "_intends to bring strong dynamic type checking_"  and "_ensures that your variables always match the model definition and validation constraints you added to them_". This is exactly what I wanted to achieve by using your lib. I wanted to be sure that all objects I use are valid but at the same time, I can work with them like with original objects but with additional validation.  The issue I described makes this library useless by default because objects wrapped by your lib makes them invalid from the start. Without your library I could do sth like this:

```
function getFoo(originalObj) {
  const copy = Object.assign({}, originalObj);
  // some validation
  copy.status = copy.status || 'pending'
  return copy
}

const foo = getFoo({ id: 1 });

console.log(foo);    // { id: 1, status: 'pending' }
console.log(foo.hasOwnProperty('id'));    // true
console.log(foo.hasOwnProperty('status')); // true
```

Even if I get additional validation and receive a copy of the original object, I expect that `hasOwnProperty('status')` will return true because this is expected default property if none was provided.

To give you some other practical example please take a look at the line from halson library which I'm using to create HAL compatible responses in my REST API.
https://github.com/seznam/halson/blob/master/index.js#L10

Because of the issue I described, the final response is missing default properties and the response is not compatible with a JSON schema.

IMHO there should be an additional trap on the proxy: `getOwnPropertyDescriptor` which should check whether a property being tested was not provided but its default is defined. If a default was specified then hasOwnProperty call should return 'true'.

**Comment by sylvainpolletvillard** [comment]
Models can validate properties no matter if they are own properties or come from a parent prototype. This is an intended feature that is required for model inheritance or type-checking on native objects for example.

The current implementation does not violate anything, you just misunderstood the documentation.`.defaults({  status: 'pending' })` is supposed to be an alias for `Foo.prototype.status = "pending"`, and properties in the prototype are by definition not `ownProperties`. Also it makes sense to me that default values are specific to the model, and not to the instance, so should not be own properties.

What you are asking me, making default values ownProperties, would imply many breaking changes:
- `delete foo.status` would resolve in `foo.status` to be `undefined` instead of the default value
- `class FooChild extends Foo { status(){ return "not a string" ) }` would be valid and break the Foo model definition, because ownProperties override class methods
- changing the default value of a model would not apply anymore to previously created model instances
- basically anyone using `hasOwnProperty` in the intended way will have breaking changes, especially if using model inheritance

So I'm not going to change the defaults behaviour, sorry. But I have an easy workaround for you : 
```
const ModelWithOwnDefaults = function(definition, defaults){
    const model = Model(definition);
    return function(data){
         data = Object.assign({}, defaults, data) // assign defaults as own props
         return new model(data)
    }
}

class Foo extends ModelWithOwnDefaults ({
  id: Number,
  status: String,
},{
  status: 'pending',
}) { }

const foo = new Foo({ id: 1 }); // init only with id

console.log(foo.hasOwnProperty('status')); // true
```

**Comment by hiredgun** [comment]
IMO using an external lib for validation should be transparent and it shouldn't force me to create some custom hacks just to achieve expected behavior. I want to work with a validated object in the same way like I would with the one created manually, with default values and without any additional library. Your arguments apply to implementation details which shouldn't affect expected, valid object. At the moment I don't have time to dive deeper into the implementation but:

- regardgin `delete foo.status` - probably it could be handled with 'deleteProperty' trap
- regarding `class FooChild extends Foo { status(){ return "not a string" ) }` - but now it also doesn't work:
```
const { Model } = require('objectmodel');

class Foo extends Model({
  id: Number,
  status: String,
}).defaults({
  status: 'pending',
}) {
  id() {
    return "blach";
  }
}
```
and try
```
const foo = new Foo({ id: 1 }); // init only with id

console.log(foo.hasOwnProperty('id'));  // true
console.log(foo.id);
delete foo.id;
console.log(foo.id); // TypeError: expecting id to be Number, got Function id
```
or
```
const foo = new Foo({ id: 1 }); // init only with id

console.log(foo.hasOwnProperty('id'));  // true
console.log(foo.id()); //  TypeError: foo.id is not a function
```
- regarding "_changing the default value ..._" - why would someone want to change default value of a model, it's a contract, it should be immutable

- regarding the last point, you said that until version 3 it was a "bug" and people were using this feature, but now it's fixed and it became a breaking change

I won't argue anymore. I choose this lib because it was the only one having a clean way of defining expected object structure (I almost got used to the lack of common assertions). I don't want to implement custom logic just to alter the behavior of the lib because it indicates that sth is wrong with the lib or it's not the lib for my use-case and also it won't make my code cleaner. Probably I will pick sth else but anyway thank you for your help.

**Comment by sylvainpolletvillard** [comment]
The code examples you posted are working the intended way. The resolution priority is:
ownProperty > class prototype property > model prototype property

Model contracts are not immutable, this is the whole point of dynamic type checking: you are able to change the model definition at anytime during execution.

The root of the problem here is that you are using `hasOwnProperty`  instead of the `in` operator. This is not only about default values, you would have the exact same issue with model inheritance or `class extends`. Whatever library you choose, you will have to either understand how prototypal inheritance works, or not use it at all.
